README.md
# cwflutter

## Compatibility

+ iOS: ConfigWise SDK flutter plugin is only supported by mobile devices:
  + with A9 or later processors.
  + iOS 14.5 and newer.
+ Android: Minimal Android 7.0 (API-24)
  + Installed Google AR Core: https://play.google.com/store/apps/details?id=com.google.ar.core
  + See full list of supported devices, here: https://developers.google.com/ar/devices

## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

To use this plugin, add `cwflutter` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

Get the [COMPANY_AUTH_TOKEN Credentials](https://manage.configwise.io) for your AR project. For more details go to [https://docs.configwise.io/api/setup-authentication-token//](https://docs.configwise.io/api/setup-authentication-token/)

Import `package:cwflutter/cwflutter.dart`, and initiate `Cwflutter Plugin` with your credentials.

### Integration

```dart
    try {
      await Cwflutter.initialize(
        defaultTargetPlatform == TargetPlatform.iOS
          ? "YOUR_IOS_COMPANY_AUTH_TOKEN"
          : "YOUR_ANDROID_COMPANY_AUTH_TOKEN",
        1 * 60 * 60,       // 1 hr
        400 * 1024 * 1024, // 400 Mb - we recommend to set 400 Mb or more for androidLowMemoryThreshold
        true
      );
    } on PlatformException catch (e) {
      print('Failed to initialize ConfigWise SDK due error: ${error.message}');
    }

    try {
      await Cwflutter.signIn();
    } on PlatformException catch (e) {
      print('Unable to pass authorization due error: ${error.message}');
    }
```

**NOTICE:** See examples (how to initialize and pass authorization) in our example project - see `_initConfigWiseSdk()` function in the 
[example/lib/main.dart](example/lib/main.dart)  

## Usage

We highly recommend to use our example project (code) to get first experience.
See: [example/lib/main.dart](example/lib/main.dart)

### Depend on it

Follow the [installation instructions](https://pub.dev/packages/cwflutter/install) from Dart Packages site.

### Update Info.plist

The plugin use native view from ARKit, which is not yet supported by default. To make it work add the following code to `Info.plist`:
```xml
    <key>io.flutter.embedded_views_preview</key>
    <string>YES</string>
```
ARKit uses the device camera, so do not forget to provide the `NSCameraUsageDescription`. You may specify it in `Info.plist` like that:
```xml
    <key>NSCameraUsageDescription</key>
    <string>Describe why your app needs AR here.</string>
``` 


CHANGELOG.md
## 1.0.0

* Initial release of ConfigWise SDK Flutter plugin.

## 1.0.1

* Loading of 3D models is implemented. We automatically put loaded models to AR scene.
* Ability to rotate, move, scale shown 3D models (using gestures).

## 1.0.4

* ConfigWise Android SDK is integrated to ConfigWise SDK Flutter plugin.
So, both mobile platforms (iOS and Android) are supported by our plugin at this moment.

## 1.0.5

* Updated iOS bridge to use latest ConfigWiseSDK v1.3.1. Xcode 12 / Swift 5.3 compatibility.
* Implemented functionality to retrieve AppContent data from ConfigWise backend.

## 1.0.6

* Gradle (in Android part) has been updated up to latest versions.

* iOS and Android bridges extended to support 'remove model from AR scene' feature and to support 'reset selected model in AR scene'.
Flutter example app (AR sceen is extended to show 'add', 'confirm' and 'delete' buttons which required to manage multiple products in the AR scene).

* Couple of defects in Android bridge have been fixed (defects about AR native code invocation of Android SDK).
Few defects are fixed in deserializers (Vector3Converter, Vector4Converter).
Flutter example app (AR sceen is extended to provide 'add new product' in to AR scene).

* 'Show Measurement' feature is implemented in iOS bridge and in the Flutter example app. NOTICE: This feature doesn't work on Android devices, 
because Android ConfigWiseSDK doesn't support it yet (this feature is available in iOS SDK only).

## 1.0.7

* Flutter plugin updated to use latest versions of ConfigWise SDKs (Android SDK v1.2.10, iOS SDK v1.3.3).
iOS example migrated to Xcode 12. Flutter AppListItemEntity and ComponentEntity have been fixed to support 
new 'Secured Downloading' mechanism.

* Optional pagination parameters (offset, max) have been added in the obtainAllComponents() and in 
the obtainAllAppListItems() functions.
NOTICE: Use null values in these params to obtain all entities in the transaction (without pagination).

* 'Secured Downloading' and 'Usage of downloading cache' (to increase loading performance) features have been 
implemented in the ConfigWise Flutter plugin.

## 1.0.8

* Flutter plugin updated to use latest version of ConfigWise iOS SDKs (v1.3.4)

* Minimal deployment target has been set as iOS 12.0

## 1.0.9

* Flutter plugin updated to use latest version of ConfigWise iOS SDKs (v1.3.8)

* Improvements in floor detection.

* Memory leaks (in iOS ConfigWiseSDK) have been fixed.

* Flutter plugin has been updated to use latest version of Android ConfigWiseSDK (v1.2.12)

* `onArPlaneDetected` callback interface has been renamed to `onArFirstPlaneDetected`.

* Two redundant callback functions have been removed from interface (`onArSessionInterrupted`, `onArSessionInterruptionEnded`).

* Extra ConfigWiseSDK initialization options have been added:

    * `dbAccessPeriod` (sec, 0 by default) - set number of seconds here if you wish 
    to query locally cached data instead to request data from server DB.
    Eg: `1hr = 1 * 60 * 60 sec`.
    Set 0 to always request data from server DB.
    
    * `lightEstimateEnabled` (true by default) - If enabled then real environment lights detected by your camera will be 
    used to calculate visualisation settings in the AR scene.

* Pagination issues (to obtain entities) have been fixed.

## 1.0.10

* Flutter plugin has been updated to use latest version of Android ConfigWiseSDK (v1.2.13).

* Obfuscator, minificator and shrinker rules (specific for Google R8 obfuscator) have been fixed in 
the Android part to exclude obfuscating of code which uses Java reflection on runtime, such as:
`org.greenrobot.eventbus.Subscribe`, `androidx.lifecycle.Lifecycle`, `androidx.lifecycle.LifecycleObserver`.

* Now, you can skip usage of `--no-shrink` build parameter in your release builds.

* Flutter plugin updated to use latest version of ConfigWise iOS SDKs (v1.3.9).

* Small bugs have been fixed in iOS ConfigWiseSDK (specific for Product Link functionality).

## 1.0.11

* Pinch gestures (to resize 3D models in the AR scene) have been disabled.

* Flutter plugin has been updated to use latest version of Android ConfigWiseSDK (v1.2.13).

* Bug has been fixed (sometimes 3D model is locked and unable to move it in AR scene in Android app).

* `onModelLoadingProgress` callback interface has been attached to Android bridge.
NOTICE: Currently (on Android platform) this callback executed twise. 1'st time - when AR starts 
loading of 3D model (`progress` parameter is `0`). 2'nd time (`progress` parameter is `100`) it will 
be run after 3D mode loading completed). Further Android `onModelLoadingProgress` functionality will 
be implemented successfully (it will be executed multiple times instead of currently twice).
On iOS platform this callback executes multiple times (`progress` parameter gives values `0...100`).
Use this callback to show (hide) progress indicator of loading.

## 1.0.12

* Flutter plugin has been updated to use latest version of Android ConfigWiseSDK (v1.2.15).

* Localization of AR error messages have been implemented in Android ConfigWiseSDK (EN, NL languages).

* Small improvements in Android ConfigWiseSDK.


## 1.0.13

* Flutter plugin has been updated to use latest version of Android ConfigWiseSDK (v1.2.16).

* Low memory management has been implemented in Android ConfigWiseSDK. This throws critical error (onArCriticalError) 
if not enough RAM to continue AR session.

## 1.0.14

* Flutter plugin has been updated to use latest version of Android ConfigWiseSDK (v1.2.18).

* Android: Runtime crash `Fatal Exception: com.google.ar.core.exceptions.FatalException` has been handled and fixed.

* Android: We fixed memory-checker and critical error handling to guarantee thread-safe.

* Android: LOW_MEMORY_THRESHOLD parameter has been made as customizable parameter. You can set it 
in `Cwflutter.initialize()` invocation (see `androidLowMemoryThreshold` parameter).

example/lib/main.dart
import 'dart:collection';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:cwflutter/cwflutter.dart';
import 'package:cwflutter/widget/ar_configuration.dart';
import 'package:cwflutter/domain/app_list_item_entity.dart';
import 'package:fluttertoast/fluttertoast.dart';

import 'ar_page.dart';

void main() {
  // See: https://flutter.dev/docs/testing/errors
  // FlutterError.onError = (FlutterErrorDetails details) {
  //   FlutterError.dumpErrorToConsole(details);
  //   if (kReleaseMode) {
  //     exit(1);
  //   }
  // };

  runApp(MaterialApp(home: MyApp()));
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';
  bool _isConfigWiseSdkInitialized = false;
  Object _error;
  AppListItemEntity _currentCategory;
  List<AppListItemEntity> _currentAppContent = List();
  final Queue<AppListItemEntity> _goBackStack = Queue();

  @override
  void initState() {
    super.initState();
    _initMyState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> _initMyState() async {
    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    _retrievePlatformVersion();
    _initConfigWiseSdk();
  }

  Future<void> _retrievePlatformVersion() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      platformVersion = await Cwflutter.platformVersion;
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  Future<void> _initConfigWiseSdk() async {
    Cwflutter.initialize(
        defaultTargetPlatform == TargetPlatform.iOS
          ? "YOUR_IOS_COMPANY_AUTH_TOKEN"
          : "YOUR_ANDROID_COMPANY_AUTH_TOKEN",
        1 * 60 * 60,       // 1 hr
        400 * 1024 * 1024, // 400 Mb - we recommend to set 400 Mb or more for androidLowMemoryThreshold
        true
    )
        .then((isInitialized) {
          if (!isInitialized) {
            return Future.value(false);
          }

          return Cwflutter.signIn();
        })
        .then((isInitialized) {
          setState(() {
            _isConfigWiseSdkInitialized = isInitialized;
            _error = null;
          });

          _retrieveAppContent(_currentCategory);
        })
        .catchError((error) {
          setState(() {
            _isConfigWiseSdkInitialized = false;
            _error = error;
          });
        });
  }

  Future<void> _retrieveAppContent(AppListItemEntity category) async {
    Cwflutter.obtainAllAppListItems(category?.id, null, null)
        .then((appListItems) {
          setState(() {
            _currentAppContent = appListItems;
          });
        })
        .catchError((error) {
          setState(() {
            _error = error;
          });
        });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('ConfigWise SDK Demo'),
      ),
      body: Container(
        margin: EdgeInsets.all(16),
        child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: ArConfiguration.values.map((configuration) => _showSupportedArConfiguration(context, configuration)).toList(),
              ),

              Text(""),

              Row(
                  children: [
                    Text('Running on:'),
                    Spacer(),
                    Text(' $_platformVersion'),
                  ]
              ),

              Text(""),

              _showConfigWiseSdkInfo(context),

              Text(""),

              _showAppContent(context),

              Spacer(),
            ]
        ),
      ),
    );
  }

  Widget _showSupportedArConfiguration(BuildContext context, ArConfiguration configuration) {
    return Row(children: [
      Text('$configuration:'),
      Spacer(),
      FutureBuilder<bool>(
          future: Cwflutter.checkConfiguration(configuration),
          builder: (context, snapshot) {
            if (!snapshot.hasData) {
              return Text(' loading');
            }
            return Text(snapshot.data ? ' supported' : ' not supported');
          })
    ]);
  }

  Widget _showConfigWiseSdkInfo(BuildContext context) {
    if (_error != null) {
      return Text(
        'ERROR: ${_error}',
        style: TextStyle(color: Colors.red),
      );
    }

    if (!_isConfigWiseSdkInitialized) {
      return CircularProgressIndicator();
    }

    return Text("ConfigWise SDK initialized.");
  }

  Widget _showAppContent(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Row(
            children: [
              IconButton(
                icon: Icon(Icons.chevron_left),
                iconSize: 36,
                color: _currentCategory == null ? Colors.grey : Colors.black,
                onPressed: () {
                  if (_currentCategory == null) { return; }
                  if (_goBackStack.isNotEmpty) {
                    final AppListItemEntity goBackCategory = _goBackStack.removeLast();
                    setState(() {
                      _currentCategory = goBackCategory;
                      _currentAppContent = List();
                    });
                    _retrieveAppContent(goBackCategory);
                  }
                },
              ),
              Spacer(),
              Text(
                  _currentCategory?.label ?? "App Content",
                style: Theme.of(context).textTheme.headline5,
              ),
              Spacer(),
            ]
        ),

        ConstrainedBox(
          constraints: new BoxConstraints(
            minHeight: 160.0,
            maxHeight: 320.0,
          ),
          child: ListView.builder(
            itemCount: _currentAppContent.length,
            shrinkWrap: true,
            itemBuilder: (context, index) {
              final appListItem = _currentAppContent[index];

              if (appListItem.type == 'MAIN_PRODUCT') {
                return AppListItemCellProduct(
                  appListItem: appListItem,
                  onTap: (appListItem) {
                    Cwflutter.obtainComponentById(appListItem.component_id)
                        .then((component) {
                          Navigator.of(context).push<void>(
                              MaterialPageRoute(builder: (c) => ArPage(initialComponent: component))
                          );
                        })
                        .catchError((error) {
                          Fluttertoast.showToast(
                              msg: 'ERROR: $error',
                              toastLength: Toast.LENGTH_LONG,
                              gravity: ToastGravity.CENTER,
                              timeInSecForIosWeb: 2
                          );
                        });
                  },
                );
              } else if (appListItem.type == 'NAVIGATION_ITEM') {
                return AppListItemCellCategory(
                  appListItem: appListItem,
                  onTap: (appListItem) {
                    _goBackStack.addLast(_currentCategory);
                    setState(() {
                      _currentCategory = appListItem;
                      _currentAppContent = List();
                    });
                    _retrieveAppContent(appListItem);
                  }
                );
              } else if (appListItem.type == 'OVERLAY_IMAGE') {
                return FutureBuilder<String>(
                    future: Cwflutter.obtainFile(appListItem.imageFileKey),
                    builder: (context, snapshot) {
                      return Image.file(
                        new File(snapshot.hasData ? snapshot.data : ''),
                        height: 100,
                        fit: BoxFit.fitWidth,
                      );
                    }
                );
              }

              return ListTile(title: null, subtitle: null);
            },
          ),
        )
      ],
    );
  }
}

typedef OnTapCallback = void Function(AppListItemEntity appListItem);

class AppListItemCellProduct extends StatelessWidget {
  const AppListItemCellProduct({Key key, this.appListItem, this.onTap}) : super(key: key);
  final AppListItemEntity appListItem;
  final OnTapCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Card(
      child: InkWell(
        onTap: () {
          onTap(appListItem);
        },
        child: ListTile(
          leading: FutureBuilder<String>(
              future: Cwflutter.obtainFile(appListItem.imageFileKey),
              builder: (context, snapshot) {
                File imageFile = new File(snapshot.hasData ? snapshot.data : '');
                if (imageFile.existsSync()) {
                  return Image.file(
                    imageFile,
                    width: 50,
                    height: 50,
                    fit: BoxFit.cover,
                  );
                } else {
                  return Icon(Icons.image, size: 50);
                }
              }
          ),
          title: Text(
            appListItem.label,
            style: Theme.of(context).textTheme.subtitle1,
          ),
          subtitle: Text(
            appListItem.description,
            style: Theme.of(context).textTheme.caption,
          ),
        ),
      ),
    );
  }
}

class AppListItemCellCategory extends StatelessWidget {
  const AppListItemCellCategory({Key key, this.appListItem, this.onTap}) : super(key: key);
  final AppListItemEntity appListItem;
  final OnTapCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Card(
      child: InkWell(
        onTap: () {
          onTap(appListItem);
        },
        child: ListTile(
          leading: FutureBuilder<String>(
              future: Cwflutter.obtainFile(appListItem.imageFileKey),
              builder: (context, snapshot) {
                return Image.file(
                  new File(snapshot.hasData ? snapshot.data : ''),
                  width: 50,
                  height: 50,
                  fit: BoxFit.cover,
                );
              }
          ),
          trailing: Icon(Icons.arrow_right),
          title: Text(
            appListItem.label,
            style: Theme.of(context).textTheme.subtitle1,
          ),
          subtitle: null,
        ),
      ),
    );
  }
}